Validation for case and coalesce clause types

romeovs opened the pull request:
Closes https://github.com/metabase/metabase/issues/57674

Adds validation for `:case` and `:if` clauses, making sure all argument positions that are returned have the same type. 

### To Verify

1. New -> Question -> Sample Database -> Orders
2. Add a custom column with case custom expression like 
   ```
   case([Discount] < 3, "cool", 3)
   ```
3. There should be an error `Type are incompatible.`

romeovs committed:
Add validation for case/if clause types

romeovs committed:
Add e2e repro for metabase#57674

romeovs committed:
Avoid passing nil default into expressions

romeovs committed:
Check for type compatibility, not equality

romeovs committed:
Allow unknoen types

romeovs committed:
Make e2e test more refined

romeovs committed:
Fix e2e test

romeovs committed:
Fix clj test

romeovs commented:
@camsaul I tagged you because you are working on the malli schema's. Maybe this PR needs to be postponed?

romeovs committed:
Fix e2e test

romeovs committed:
Handle set arguments to compatible?

romeovs committed:
Fix docstring

ranquild reviewed (dismissed):
I think this is not correct. `type/Date` and `type/Time` have `type/Temporal` as the parent time, but it doesn't mean that you could mix them in a single expression. There was some different function to check whether expressions of different types are compatible AFAIK. I also remember that we special-case `type/Date` mixed with `type/DateTime` in several places.

romeovs committed:
Use assignable? instead of most-common-ancestor

romeovs committed:
Allow Date/DateTime/... to mix

romeovs committed:
Merge branch 'master' into 57674-validation-for-case-clause-types

romeovs committed:
Use best-return-type to determine if all types are compatible

romeovs committed:
Handle falsy default

romeovs committed:
Merge branch 'master' into 57674-validation-for-case-clause-types

romeovs committed:
Add missing map

romeovs commented:
@ranquild I've switched the type check to use the same logic that Cam just introduced, please have another look.

camsaul commented on src/metabase/lib/schema/expression/conditional.cljc:
```diff
@@ -75,18 +75,39 @@
+                                          {:error/message "Valid [pred expr] pair"}
+                                          #_pred [:ref ::expression/boolean]
+                                          #_expr [:ref ::expression/expression]]]]
+                                       [:default [:? [:schema [:ref ::expression/expression]]]]))
+      [:fn
+       ;; further constrain this so all of the exprs are of the same type
+       {:error/message "All clauses should have the same type"}
+       (fn
+         [[_tag _opts pred-expr-pairs default]]
+         (let [exprs (concat
+                      (map second pred-expr-pairs)
+                      (when (some? default)
+                        [default]))
+               types (map expression/type-of exprs)
+               return-type (case-coalesce-return-type types)]
+           (or
+            (not= return-type :type/*)
+            (some #{:expression/type.unknown} types))))]]])
+
   (defmethod expression/type-of-method tag
     [[_tag _opts pred-expr-pairs default]]
     (let [exprs (concat
                  (map second pred-expr-pairs)
-                 (when default
+                 (when (some? default)
```
good catch, can you add a test for this?

romeovs replied:
As of the changes in this branch, I don't think it's possible anymore to hit a case where this would be triggered since you'd need at least one predicate-expression pair and a default value.

The bug was basically only trigged when `default` is `false`, but because of the new schema check that mean the predicate-expression pairs have to return `:type/Boolean` or `::expression/type.unknown` to be valid. Omitting `default` in the doesn't do anything to change the type anymore.

I still think it's valid to leave the `some?` check in for completeness (and symmetry with the schema check).

camsaul reviewed (changes requested):
- I see you fixed `:case` with a `false` default value, can you add a test for that?
- We should constrain `:coalesce` as well using the same rules.
- I think you accidentally disabled the original schema when you wrapped it in `(:schema ...)` -- we should probably make sure we have tests that verify the schema fails when we don't match that shape
- I'm not convinced this PR really does what it claims to do -- if I understand this correctly this doesn't prevent someone from using two incompatible types as long as they're distantly related, for example `DATE` and `TIME` (which are both `:type/Temporal`).  I don't think we really have a good notion of whether two types are compatible or not, so maybe we should think about adding something like that. Either way, this type checking is still an improvement over the old stuff so I don't want to let the perfect be the enemy of the good if you can't figure something out.

romeovs committed:
Update comment to clarify compatibility omission

romeovs committed:
Unwrap schema

romeovs committed:
Add test for malformed case expressions

romeovs committed:
Add type check to coalaesce too

romeovs committed:
Add check to coalesce too

romeovs commented:
> I see you fixed :case with a false default value, can you add a test for that?

As per [my comment](https://github.com/metabase/metabase/pull/59038#discussion_r2153997926), it's now also impossible to reach this case as it would not validate.

> We should constrain :coalesce as well using the same rules.
Done!

> I think you accidentally disabled the original schema when you wrapped it in (:schema ...) -- we should probably make sure we have tests that verify the schema fails when we don't match that shape
Oof! Good one. Fixed it and added a test.

> I'm not convinced this PR really does what it claims to do -- if I understand this correctly this doesn't prevent someone from using two incompatible types as long as they're distantly related, for example DATE and TIME (which are both :type/Temporal). I don't think we really have a good notion of whether two types are compatible or not, so maybe we should think about adding something like that. Either way, this type checking is still an improvement over the old stuff so I don't want to let the perfect be the enemy of the good if you can't figure something out.
You're right: it's a best-effort approach to restrict the usage of `case` and `coalesce`. I don't think we have anything that does full type compatibility yet, or if it's even possible to create something like it, but I think this is an improvement over not doing anything!

ranquild reviewed (commented):
@romeovs please write a test for `:default false` as @camsaul  suggested. Also add tests for cases that the most common type would not cover, like `Date` + `DateTime`

romeovs committed:
Add tests for :default false

romeovs committed:
Add tests for temporal types

ranquild commented on e2e/test/scenarios/custom-column/custom-column-reproductions.cy.spec.js:
```diff
@@ -2056,3 +2059,86 @@ describe("Issue 58230", () => {
   });
 });
+
+describe("issue 57674", () => {
```
I don't think we need to enumerate possible cases with an e2e test. Either no e2e test, or just one case

ranquild reviewed (commented):
Left a comment about the FE changes.

romeovs committed:
Move combination tests to BE test

metamben commented on src/metabase/lib/schema/expression/conditional.cljc:
```diff
@@ -75,26 +75,63 @@
+       {:error/message "All clauses should have the same type"}
+       (fn
+         [[_tag _opts pred-expr-pairs default]]
+         (let [exprs (concat
+                      (map second pred-expr-pairs)
+                      (when (some? default)
+                        [default]))
```
You could `(cond->> (map ...) (some? default) (cons default))`, since we don't care about the  order.

romeovs replied:
I'll keep it as is for consistency with the `expression/type-of-method` function below, where the order _does_ matter.

metamben commented on src/metabase/lib/schema/expression/conditional.cljc:
```diff
@@ -75,26 +75,63 @@
+               return-type (case-coalesce-return-type types)]
+           (or
+            (not= return-type :type/*)
+            (some #{:expression/type.unknown} types))))]]])
```
We don't even need to calculate `return-type` in this case, do we?

metamben commented on src/metabase/lib/schema/expression/conditional.cljc:
```diff
@@ -75,26 +75,63 @@
+             return-type (case-coalesce-return-type types)]
+         (or
+          (not= return-type :type/*)
+          (some #{:expression/type.unknown} types))))]]])
```
here too.

metamben reviewed (approved):
LGTM

romeovs committed:
Use separate deftests for valid and invalid case and coalesce tests

romeovs committed:
Fold repetitive parts into are expression

romeovs committed:
Merge branch 'master' into 57674-validation-for-case-clause-types

camsaul commented on test/metabase/lib/schema/expression/conditional_test.cljc:
```diff
@@ -32,6 +32,97 @@
+      [[[true true]] true]
+
+      [[[true true]] true])
+
+    (are [args] (false?
+                 (mr/validate :mbql.clause/case
+                              (into [:case {:lib/uuid (str (random-uuid))}] args)))
+      [1]
+
+      [[] 1])))
```
```suggestion
    (are [args] (false?
                 (mr/validate :mbql.clause/case
                              (into [:case {:lib/uuid (str (random-uuid))}] args)))
      [1]
      [[] 1])))
```

camsaul commented on test/metabase/lib/schema/expression/conditional_test.cljc:
```diff
@@ -70,6 +161,110 @@
+      [(value-expr :type/Date "2023-03-08")
+       (value-expr :type/Time "15:03:55")])
+
+    (are [args] (false?
```
+1 on what Braden said, can you Can you split this out into a separate test?

camsaul reviewed (approved):
I'm going to ✔️ this so you don't have to wait for another re-review but I left some notes about splitting up the tests a little and formatting them more idiomatically you should address first... also Braden left some similar comments last week I don't think have been addressed yet

romeovs committed:
Formatting of test

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

romeovs committed:
Formatting of test

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

romeovs committed:
Formatting of test

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

romeovs committed:
Split up more tests

romeovs committed:
Formatting in tests

romeovs committed:
Merge branch 'master' into 57674-validation-for-case-clause-types

romeovs committed:
Merge branch 'master' into 57674-validation-for-case-clause-types
